File: src/main/java/zm/unza/counseling/service/RiskAssessmentService.java
package zm.unza.counseling.service;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import zm.unza.counseling.dto.request.RiskAssessmentRequest;
import zm.unza.counseling.entity.Client;
import zm.unza.counseling.entity.RiskAssessment;
import zm.unza.counseling.repository.ClientRepository;
import zm.unza.counseling.repository.RiskAssessmentRepository;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Service interface for risk assessment management
 */
public interface RiskAssessmentService {
    
    /**
     * Get all risk assessments with pagination
     * @param pageable pagination information
     * @return paginated list of risk assessments
     */
    Page<RiskAssessment> getAllRiskAssessments(Pageable pageable);

    /**
     * Get risk assessment by ID
     * @param id the risk assessment ID
     * @return the risk assessment
     */
    RiskAssessment getRiskAssessmentById(Long id);

    /**
     * Create risk assessment
     * @param request the risk assessment request
     * @return the created risk assessment
     */
    RiskAssessment createRiskAssessment(RiskAssessmentRequest request);

    /**
     * Update risk assessment
     * @param id the risk assessment ID
     * @param request the risk assessment request
     * @return the updated risk assessment
     */
    RiskAssessment updateRiskAssessment(Long id, RiskAssessmentRequest request);

    /**
     * Delete risk assessment
     * @param id the risk assessment ID
     */
    void deleteRiskAssessment(Long id);

    /**
     * Get risk assessments by client
     * @param clientId the client ID
     * @return list of risk assessments for the client
     */
    List<RiskAssessment> getRiskAssessmentsByClient(Long clientId);

    /**
     * Get assessments for a specific client
     * @param clientId the client ID
     * @return list of risk assessments for the client
     */
    List<RiskAssessment> getAssessmentsForClient(Long clientId);

    /**
     * Get high risk assessments
     * @return list of high risk assessments
     */
    List<RiskAssessment> getHighRiskAssessments();

    /**
     * Get risk assessment stats
     * @return risk assessment statistics
     */
    Object getRiskAssessmentStats();

    /**
     * Escalate risk assessment
     * @param id the risk assessment ID
     * @return the escalated risk assessment
     */
    RiskAssessment escalateRiskAssessment(Long id);

    /**
     * Export risk assessment data
     * @param format the export format
     * @return the exported risk assessment data
     */
    byte[] exportRiskAssessmentData(String format);

    /**
     * Get latest risk assessment for client
     * @param clientId the client ID
     * @return the latest risk assessment
     */
    RiskAssessment getLatestRiskAssessmentForClient(Long clientId);

    /**
     * Get risk assessment trend for client
     * @param clientId the client ID
     * @return the risk assessment trend data
     */
    Object getRiskAssessmentTrend(Long clientId);

    /**
     * Get risk assessment summary
     * @return the risk assessment summary
     */
    Object getRiskAssessmentSummary();

    /**
     * Get assessments requiring follow-up
     * @return list of assessments requiring follow-up
     */
    List<RiskAssessment> getAssessmentsRequiringFollowUp();

    /**
     * Get risk assessments by assessor
     * @param assessorId the assessor ID
     * @return list of risk assessments by assessor
     */
    Object getRiskAssessmentsByAssessor(Long assessorId);

    /**
     * Get risk assessment analytics
     * @return the risk assessment analytics
     */
    Object getRiskAssessmentAnalytics();
}

File: src/services/riskAssessmentService.ts
import api from './api';
import { PaginatedResponse } from '../types/api';

/**
 * Risk Assessment Type Definitions
 */
export interface RiskAssessment {
  id: string;
  client: {
    id: string;
    firstName: string;
    lastName: string;
    email: string;
  };
  assessmentType: string;
  riskLevel: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  score: number;
  factors: string[];
  recommendations: string[];
  assessedBy: {
    id: string;
    firstName: string;
    lastName: string;
  };
  assessmentDate: string;
  notes?: string;
  escalated: boolean;
  followUpRequired: boolean;
  followUpDate?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateRiskAssessmentRequest {
  clientId: number;
  assessmentType: string;
  riskLevel: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  score: number;
  factors: string[];
  recommendations: string[];
  notes?: string;
}

export interface UpdateRiskAssessmentRequest {
  assessmentType?: string;
  riskLevel?: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  score?: number;
  factors?: string[];
  recommendations?: string[];
  notes?: string;
  escalated?: boolean;
  followUpRequired?: boolean;
  followUpDate?: string;
}

/**
 * Risk Assessment Service - Handles all risk assessment API operations
 * Required Roles: ADMIN, COUNSELOR
 */
export const riskAssessmentService = {
  /**
   * Get all risk assessments (paginated)
   * GET /api/v1/risk-assessments?page=0&size=10
   */
  getAllRiskAssessments: async (page = 0, size = 10): Promise<PaginatedResponse<RiskAssessment>> => {
    const response = await api.get(`/risk-assessments?page=${page}&size=${size}`);
    return response.data.data;
  },

  /**
   * Get risk assessment by ID
   * GET /api/v1/risk-assessments/{id}
   */
  getRiskAssessmentById: async (id: string): Promise<RiskAssessment> => {
    const response = await api.get(`/risk-assessments/${id}`);
    return response.data.data;
  },

  /**
   * Create risk assessment
   * POST /api/v1/risk-assessments
   */
  createRiskAssessment: async (data: CreateRiskAssessmentRequest): Promise<RiskAssessment> => {
    const response = await api.post('/risk-assessments', data);
    return response.data.data;
  },

  /**
   * Update risk assessment
   * PUT /api/v1/risk-assessments/{id}
   */
  updateRiskAssessment: async (id: string, data: UpdateRiskAssessmentRequest): Promise<RiskAssessment> => {
    const response = await api.put(`/risk-assessments/${id}`, data);
    return response.data.data;
  },

  /**
   * Delete risk assessment (Admin only)
   * DELETE /api/v1/risk-assessments/{id}
   */
  deleteRiskAssessment: async (id: string): Promise<void> => {
    await api.delete(`/risk-assessments/${id}`);
  },

  /**
   * Get risk assessments by client
   * GET /api/v1/risk-assessments/client/{clientId}
   */
  getRiskAssessmentsByClient: async (clientId: string): Promise<RiskAssessment[]> => {
    const response = await api.get(`/risk-assessments/client/${clientId}`);
    return response.data.data;
  },

  /**
   * Get high risk assessments
   * GET /api/v1/risk-assessments/high-risk
   */
  getHighRiskAssessments: async (): Promise<RiskAssessment[]> => {
    const response = await api.get('/risk-assessments/high-risk');
    return response.data.data;
  },

  /**
   * Get risk assessment statistics
   * GET /api/v1/risk-assessments/stats
   */
  getStats: async () => {
    const response = await api.get('/risk-assessments/stats');
    return response.data.data;
  },

  /**
   * Escalate risk assessment
   * POST /api/v1/risk-assessments/{id}/escalate
   */
  escalateRiskAssessment: async (id: string): Promise<RiskAssessment> => {
    const response = await api.post(`/risk-assessments/${id}/escalate`);
    return response.data.data;
  },

  /**
   * Export risk assessments (Admin only)
   * GET /api/v1/risk-assessments/export?format=csv
   */
  exportRiskAssessments: async (format: 'csv' | 'pdf' = 'csv'): Promise<Blob> => {
    const response = await api.get(`/risk-assessments/export?format=${format}`, {
      responseType: 'blob'
    });
    return response.data;
  },

  /**
   * Get latest risk assessment for client
   * GET /api/v1/risk-assessments/client/{clientId}/latest
   */
  getLatestForClient: async (clientId: string): Promise<RiskAssessment> => {
    const response = await api.get(`/risk-assessments/client/${clientId}/latest`);
    return response.data.data;
  },

  /**
   * Get risk assessment trend for client
   * GET /api/v1/risk-assessments/client/{clientId}/trend
   */
  getTrend: async (clientId: string) => {
    const response = await api.get(`/risk-assessments/client/${clientId}/trend`);
    return response.data.data;
  },

  /**
   * Get risk assessment summary
   * GET /api/v1/risk-assessments/summary
   */
  getSummary: async () => {
    const response = await api.get('/risk-assessments/summary');
    return response.data.data;
  },

  /**
   * Get assessments requiring follow-up
   * GET /api/v1/risk-assessments/follow-up-required
   */
  getFollowUpRequired: async (): Promise<RiskAssessment[]> => {
    const response = await api.get('/risk-assessments/follow-up-required');
    return response.data.data;
  },

  /**
   * Get assessments by assessor
   * GET /api/v1/risk-assessments/assessor?assessorId={id}
   */
  getByAssessor: async (assessorId: string) => {
    const response = await api.get(`/risk-assessments/assessor?assessorId=${assessorId}`);
    return response.data.data;
  },

  /**
   * Get risk assessment analytics
   * GET /api/v1/risk-assessments/analytics
   */
  getAnalytics: async () => {
    const response = await api.get('/risk-assessments/analytics');
    return response.data.data;
  }
};
